File: java/src/main/java/com/example/gemini/FunctionCalling.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.gemini;

import com.google.genai.Client;
import com.google.genai.types.FunctionCallingConfig;
import com.google.genai.types.FunctionDeclaration;
import com.google.genai.types.GenerateContentConfig;
import com.google.genai.types.GenerateContentResponse;
import com.google.genai.types.Schema;
import com.google.genai.types.Tool;
import com.google.genai.types.ToolConfig;
import org.apache.http.HttpException;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.function.BiFunction;

public class FunctionCalling {
    public static Double functionCalling() {
        // [START function_calling]
        Client client = new Client();

        FunctionDeclaration addFunction =
                FunctionDeclaration.builder()
                        .name("addNumbers")
                        .parameters(
                                Schema.builder()
                                        .type("object")
                                        .properties(Map.of(
                                                "firstParam", Schema.builder().type("number").description("First number").build(),
                                                "secondParam", Schema.builder().type("number").description("Second number").build()))
                                        .required(Arrays.asList("firstParam", "secondParam"))
                                        .build())
                        .build();

        FunctionDeclaration subtractFunction =
                FunctionDeclaration.builder()
                        .name("subtractNumbers")
                        .parameters(
                                Schema.builder()
                                        .type("object")
                                        .properties(Map.of(
                                                "firstParam", Schema.builder().type("number").description("First number").build(),
                                                "secondParam", Schema.builder().type("number").description("Second number").build()))
                                        .required(Arrays.asList("firstParam", "secondParam"))
                                        .build())
                        .build();

        FunctionDeclaration multiplyFunction =
                FunctionDeclaration.builder()
                        .name("multiplyNumbers")
                        .parameters(
                                Schema.builder()
                                        .type("object")
                                        .properties(Map.of(
                                                "firstParam", Schema.builder().type("number").description("First number").build(),
                                                "secondParam", Schema.builder().type("number").description("Second number").build()))
                                        .required(Arrays.asList("firstParam", "secondParam"))
                                        .build())
                        .build();

        FunctionDeclaration divideFunction =
                FunctionDeclaration.builder()
                        .name("divideNumbers")
                        .parameters(
                                Schema.builder()
                                        .type("object")
                                        .properties(Map.of(
                                                "firstParam", Schema.builder().type("number").description("First number").build(),
                                                "secondParam", Schema.builder().type("number").description("Second number").build()))
                                        .required(Arrays.asList("firstParam", "secondParam"))
                                        .build())
                        .build();

        GenerateContentConfig config = GenerateContentConfig.builder()
                .toolConfig(ToolConfig.builder().functionCallingConfig(
                        FunctionCallingConfig.builder().mode("ANY").build()
                ).build())
                .tools(
                        Collections.singletonList(
                                Tool.builder().functionDeclarations(
                                        Arrays.asList(
                                                addFunction,
                                                subtractFunction,
                                                divideFunction,
                                                multiplyFunction
                                        )
                                ).build()

                        )
                )
                .build();

        GenerateContentResponse response =
                client.models.generateContent(
                        "gemini-3.8-flash",
                        "I have 57 cats, each owns 44 mittens, how many mittens is that in total?",
                        config);


        if (response.functionCalls() == null || response.functionCalls().isEmpty()) {
            System.err.println("No function call received");
            return null;
        }

        var functionCall = response.functionCalls().getFirst();
        String functionName = functionCall.name().get();
        var arguments = functionCall.args();

        Map<String, BiFunction<Double, Double, Double>> functionMapping = new HashMap<>();
        functionMapping.put("addNumbers", (a, b) -> a + b);
        functionMapping.put("subtractNumbers", (a, b) -> a - b);
        functionMapping.put("multiplyNumbers", (a, b) -> a * b);
        functionMapping.put("divideNumbers", (a, b) -> b != 0 ? a / b : Double.NaN);

        BiFunction<Double, Double, Double> function = functionMapping.get(functionName);

        Number firstParam = (Number) arguments.get().get("firstParam");
        Number secondParam = (Number) arguments.get().get("secondParam");
        Double result = function.apply(firstParam.doubleValue(), secondParam.doubleValue());

        System.out.println(result);
        // [END function_calling]
        return result;
    }
}

File: python/function_calling.py
# -*- coding: utf-8 -*-
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from absl.testing import absltest


class UnitTests(absltest.TestCase):

    def test_function_calling(self):
        # [START function_calling]
        from google import genai
        from google.genai import types

        client = genai.Client()

        def add(a: float, b: float) -> float:
            """returns a + b."""
            return a + b

        def subtract(a: float, b: float) -> float:
            """returns a - b."""
            return a - b

        def multiply(a: float, b: float) -> float:
            """returns a * b."""
            return a * b

        def divide(a: float, b: float) -> float:
            """returns a / b."""
            return a / b

        # Create a chat session; function calling (via tools) is enabled in the config.
        chat = client.chats.create(
            model="gemini-3.8-flash",
            config=types.GenerateContentConfig(tools=[add, subtract, multiply, divide]),
        )
        response = chat.send_message(
            message="I have 57 cats, each owns 44 mittens, how many mittens is that in total?"
        )
        print(response.text)
        # [END function_calling]


if __name__ == "__main__":
    absltest.main()
